File: Tutorial/Ml/Nlp/TutorialNlpByteEmpyrealmAg.cpp
// ═══════════════════════════════════════════════════════════════════════════
// OA Tutorial — Empyrealm Core Ag (Autograd Fidelity)
//
// OaEmpyrealmCore with UseFused=false: traceable Preprocess + EmpyrealmSiso
// + out_proj MatMul. Validates full Mamba-3 mixer autograd (in_proj, dt/B/C
// bias, D, norm_weight, out_proj) through the flat residual + CE head.
//
// Siblings:
//   TutorialNlpByteMamba3Ag    — untouched OaMamba3Module + Ssm/Mamba3/ reference
//   TutorialNlpByteRnnAg / TutorialNlpByteGruAg — recurrent baselines
// ═══════════════════════════════════════════════════════════════════════════

#include "OaTest.h"
#include "TutorialMl.h"
#include "TutorialNlpCommon.h"
#include <Oa/Ml.h>
#include <Oa/Ml/Byte.h>
#include <Oa/Ml/Autograd.h>
#include <Oa/Ml/OptimUtil.h>
#include <cstdlib>
#include <Oa/Runtime/Context.h>
#include <Oa/Runtime/Engine.h>
#include <Oa/Runtime/GpuTimer.h>
#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstring>
#include <vector>

static bool TutorialUseMuonOptimizer() {
	const char* env = std::getenv("OA_USE_MUON");
	return env && env[0] == '1';
}

static OaUniquePtr<OaOptimizer> TutorialMakeOptimizer(OaModule& InModel, OaF32 InLr) {
	if (TutorialUseMuonOptimizer()) {
		OaMuonAdamWConfig cfg{};
		cfg.AdamWLr = InLr;
		return MakeMuonAdamWOptimizer(InModel, cfg);
	}
	auto params = InModel.AllParameterPtrs();
	return OaMakeUniquePtr<OaAdamW>(params, InLr);
}

static constexpr OaI32 kVocabSize  = OA_BYTE_VOCAB_SIZE;  // 256 — byte vocab family
// kContextLen / kDModel come from TutorialNlpCommon.h (shared suite dims).
static constexpr OaI32 kDState     = 32;
static constexpr OaI32 kExpand     = 2;
static constexpr OaI32 kHeadDim    = 16;

class OaEmpyrealmByteLMAg : public OaModule {
public:
	OaEmpyrealmByteLMAg() {
		Core_ = OaMakeSharedPtr<OaEmpyrealmCore>(kVocabSize, kDModel,
			kDState, kExpand, kHeadDim,
			/*NGroups*/ 1, /*RopeFraction*/ 0.5f, /*Mimo*/ false, /*MimoRank*/ 1,
			/*DtMin*/ 0.001f, /*DtMax*/ 0.1f, /*DtInitFloor*/ 1e-4f, /*AFloor*/ 1e-4f,
			/*OutprojNorm*/ true);
		Head_ = OaMakeSharedPtr<OaLinear>(kDModel, kVocabSize);
		RegisterModule("core", Core_);
		RegisterModule("head", Head_);
	}

	OaMatrix Forward(const OaMatrix& InTokens) override {
		auto mixed = Core_->Forward(InTokens);
		return Head_->Forward(mixed);
	}

	[[nodiscard]] OaSharedPtr<OaEmpyrealmCore> Core() const { return Core_; }

private:
	OaSharedPtr<OaEmpyrealmCore> Core_;
	OaSharedPtr<OaLinear>        Head_;
};

static float ParamGradL1(const OaMatrix& g) {
	if (g.IsEmpty() || g.NumElements() == 0) return 0.0F;
	auto s = OaFnMatrix::Sum(OaFnMatrix::Abs(g.Reshape(OaMatrixShape{g.NumElements()})), 0);
	return s.At(0);
}

TEST(TutorialNlpByteEmpyrealmAg, TraceableAutograd) {
	printf("\n╔══════════════════════════════════════════════════════════════════╗\n");
	printf("║  OA Tutorial — Empyrealm Core Ag (Autograd Fidelity)             ║\n");
	printf("║  EmpyrealmPreprocess + EmpyrealmSiso traced path                 ║\n");
	printf("╚══════════════════════════════════════════════════════════════════╝\n\n");

	auto  model  = OaMakeSharedPtr<OaEmpyrealmByteLMAg>();
	auto  params = model->AllParameterPtrs();
	// lr 0.003: with correct [B,S,D] batching (EmpyrealmCore::Forward) the model
	// converges fast; 0.01 is hot enough to diverge late, same as the Mamba3 ref.
	auto  opt    = TutorialMakeOptimizer(*model, 0.003F);

	printf("Path: traceable split (Preprocess MatMul + EmpyrealmSiso + out_proj)\n");
	printf("Params: %lld    Optimizer: %s(lr=0.003)\n\n",
		static_cast<long long>(model->NumParameters()),
		TutorialUseMuonOptimizer() ? "Muon+AdamW" : "AdamW");

	constexpr OaI32 kSteps = 300;
	constexpr OaI32 kBatch = 64;
	NlpAllPositionSampler sampler(NlpCorpus(), kBatch);

	TutorialTrainingLoop training(*opt, OaItTrainingConfig{
		.TotalSteps     = kSteps,
		.EpochSteps     = {},
#if defined(_WIN32)
		// Same long SSM graph as the Mamba3 reference (UseFused=false delegates to
		// the identical OaMamba3Module::Forward): the Windows NVIDIA driver loses
		// the device with more than one training submission in flight, so sync
		// every step.
#endif
		.BatchSize      = kBatch,
		.SequenceLength = kContextLen,
		.SequenceUnit   = "token",
		.SourceUnitsPerSample = kContextLen,
		.SourceUnit     = "byte",
		.TimerName      = "empyrealm_ag_step",
		.Callbacks      = {}
	});

	// Ring-buffered batches: OaItTraining may submit up to MaxAsyncSubmissions()
	// steps before syncing; reusing one batchX/batchY would let the GPU read
	// overwritten token buffers while earlier steps are still in flight.
	auto& ctx = OaContext::GetDefault();
	OaVec<OaMatrix> xRing(static_cast<OaI64>(ctx.MaxAsyncSubmissions()));
	OaVec<OaMatrix> yRing(static_cast<OaI64>(ctx.MaxAsyncSubmissions()));
	OaF32 initialLoss = 0.0F;
	OaF32 lastLoss    = 0.0F;
	float inProjGradL1 = 0.0F;
	float outProjGradL1 = 0.0F;

	while (not training.Loop.IsDone()) {
		const OaI64 step = training.Loop.Index();
		OaMatrix& batchX = xRing[(step - 1) % xRing.Size()];
		OaMatrix& batchY = yRing[(step - 1) % yRing.Size()];
		sampler.NextBatch(batchX, batchY);
		opt->ZeroGrad();
		OaGradientTape tape;
		auto logits = model->Forward(batchX);
		auto loss   = OaFnLoss::CrossEntropy(logits,
			batchY.Reshape(OaMatrixShape{batchY.Size(0) * batchY.Size(1)}));
		tape.Backward(loss);
		training.Loop.Next(loss);

		if (training.Loop.Index() == 1) {
			initialLoss = training.Loop.LiveLoss();
			printf("\n─── Step-1 gradient magnitudes (traceable path, L1) ───\n");
			auto& ctx2 = OaContext::GetDefault();
			struct MagEntry { const char* name; OaMatrix result; OaI64 numel; };
			std::vector<MagEntry> entries;
			for (auto* p : params) {
				auto g = p->Data.GradMatrix();
				OaMatrix s;
				OaI64 numel = 0;
				if (!g.IsEmpty() && g.NumElements() > 0) {
					auto flat = g.Reshape(OaMatrixShape{g.NumElements()});
					auto absg = OaFnMatrix::Abs(flat);
					s = OaFnMatrix::Sum(absg, 0);
					numel = p->Data.NumElements();
				}
				entries.push_back({p->Name.CStr(), std::move(s), numel});
			}
			(void)ctx2.Execute();
			(void)ctx2.Sync();
			for (const auto& e : entries) {
				float mag = 0.0F;
				if (e.result.NumElements() > 0) mag = e.result.At(0);
				printf("  %-32s  L1=%.6f  (numel=%lld)\n",
					e.name, mag, static_cast<long long>(e.numel));
				if (std::strcmp(e.name, "in_proj") == 0) inProjGradL1 = mag;
				if (std::strcmp(e.name, "out_proj") == 0) outProjGradL1 = mag;
			}
			printf("\n");
			fflush(stdout);
		}
	}

	ASSERT_TRUE(training.Loop.Finish().IsOk());
	lastLoss = training.Loop.LastLoss();

	OaMatrix& evalX = xRing[(kSteps - 1) % xRing.Size()];
	OaMatrix& evalY = yRing[(kSteps - 1) % yRing.Size()];
	const OaF32 finalBatchAcc = NlpAccuracyAllPositions(*model, evalX, evalY, kVocabSize);
	printf("\nEvaluation:\n");
	printf("  Random-loss baseline ln(%d) = %.4f\n",
		kVocabSize, std::log(static_cast<double>(kVocabSize)));
	printf("  Bits/byte: %.4f\n", NlpBitsPerByte(lastLoss));
	printf("  Accuracy: %.1f%%\n", finalBatchAcc);

	printf("\nGeneration:\n");
	printf("  Prompt: '%s'\n", kNlpGenerationPrompt);
	printf("  Generated: '%s'\n\n", NlpGenerateGreedy(*model, kNlpGenerationPrompt,
		kNlpGenerationBytes, kVocabSize).c_str());

	EXPECT_LT(lastLoss, initialLoss);
	EXPECT_GT(finalBatchAcc, 30.0F);
	EXPECT_GT(inProjGradL1, 0.0F)  << "mixer.in_proj must receive gradient on traceable path";
	EXPECT_GT(outProjGradL1, 0.0F) << "mixer.out_proj must receive gradient on traceable path";

	const OaString ckptPath = "/tmp/empyrealm_ag.oam";
	ASSERT_TRUE(model->Save(ckptPath, *opt).IsOk());
	auto reloaded = OaMakeSharedPtr<OaEmpyrealmByteLMAg>();
	auto reloadedOpt = TutorialMakeOptimizer(*reloaded, 0.003F);
	ASSERT_TRUE(reloaded->Load(ckptPath, *reloadedOpt).IsOk());
	EXPECT_NEAR(NlpAccuracyAllPositions(*reloaded, evalX, evalY, kVocabSize), finalBatchAcc, 1.0F);
}

File: Tutorial/Py/Ml/Nlp/TutorialNlpByteEmpyrealmAg.py
#!/usr/bin/env python3
"""1:1 Python entry point for TutorialNlpByteEmpyrealmAg.cpp.

Empyrealm remains the sixteenth experimental regression member outside the
canonical 3-tokenizer × 5-architecture matrix.
"""

from __future__ import annotations

import math

# pyright: reportWildcardImportFromLibrary=false
from oa import *
import _nlp_common as nlp


class ByteEmpyrealmLM:
	def __init__(self) -> None:
		self.core = OaEmpyrealmCore(
			256,
			OaNlpSuiteModelWidth,
			DState=32,
			Expand=2,
			HeadDim=16,
			NGroups=1,
			RopeFraction=0.5,
			Mimo=False,
			MimoRank=1,
			DtMin=0.001,
			DtMax=0.1,
			DtInitFloor=1e-4,
			AFloor=1e-4,
			OutprojNorm=True,
		)
		self.head = OaLinear(OaNlpSuiteModelWidth, 256)
		for parameter in self.parameters():
			parameter.Data.SetRequiresGrad(True)

	def parameters(self):
		return (
			list(self.core.AllParameterPtrs())
			+ list(self.head.AllParameterPtrs())
		)

	def Forward(self, tokens):
		return self.head.Forward(self.core.Forward(tokens))

	def Save(self, prefix) -> None:
		self.core.Save(prefix.String() + ".core.oam")
		self.head.Save(prefix.String() + ".head.oam")

	def Load(self, prefix) -> None:
		self.core.Load(prefix.String() + ".core.oam")
		self.head.Load(prefix.String() + ".head.oam")


def main() -> None:
	if not OaInitComputeEngine():
		raise RuntimeError("OA compute engine initialization failed")

	steps = nlp.configured_steps()
	batch = nlp.configured_batch()
	generation_units = nlp.configured_generation_units()
	full_gate = (
		steps == OaNlpSuiteTrainingSteps
		and batch == OaNlpSuiteBatchSize
		and generation_units == OaNlpSuiteGenerationSourceUnits
	)
	# The recipe supplies the canonical Byte sampler. Empyrealm owns a distinct
	# model but otherwise uses the same workload as the 15 controlled members.
	recipe = OaNlpSuiteRecipe(
		OaNlpArchitecture.Gru,
		OaNlpTokenizerKind.Byte,
	)
	OaFnMatrix.SetRngSeed(OaNlpSuiteRngSeed)
	model = ByteEmpyrealmLM()
	parameters = model.parameters()
	optimizer = OaAdamW(parameters, 0.003)
	sampler = OaNlpSuiteSampler(recipe, batch)

	print("\n" + "=" * 72)
	print("  OA Python Tutorial — Byte Empyrealm · all-position LM")
	print("=" * 72)
	print(
		"Model: Byte embedding + Empyrealm SSM(state=32) + residual "
		"-> Linear(256)"
	)
	print(
		f"Params: {sum(p.Data.NumElements() for p in parameters)} · "
		"AdamW(lr=0.003)"
	)
	print(
		f"Training: {steps} steps · batch={batch} · "
		f"sequence={recipe.ContextLength()} byte tokens"
	)

	config = OaItTrainingConfig()
	config.TotalSteps = steps
	config.BatchSize = batch
	config.SequenceLength = recipe.ContextLength()
	config.SequenceUnit = "token"
	config.SourceUnit = "byte"
	config.TimerName = "empyrealm_ag_step"
	training = OaItTraining(optimizer, config)

	initial_loss = 0.0
	x = y = None
	while not training.IsDone():
		x, y = sampler.Next()
		training.RecordSourceUnits(sampler.LastSourceUnits())
		optimizer.ZeroGrad()
		with OaGradientTape() as tape:
			logits = model.Forward(x)
			loss = OaFnLoss.CrossEntropy(
				logits,
				y.Reshape([y.NumElements()]),
			)
			tape.Backward(loss)
		training.Next(loss)
		if training.Index() == 1:
			initial_loss = training.LastLoss()
	training.Finish()

	if x is None or y is None:
		raise RuntimeError("Empyrealm training produced no batch")
	final_loss = training.LastLoss()
	accuracy = 100.0 * OaFnMetric.Accuracy(model.Forward(x), y)
	generated = nlp._generate_sliding(
		model,
		OaNlpSuiteSampler(recipe, 1),
		recipe,
		generation_units,
	)

	checkpoint = OaPaths.Temp() / "oa_py_nlp_byte_empyrealm"
	model.Save(checkpoint)
	reloaded = ByteEmpyrealmLM()
	reloaded.Load(checkpoint)
	reloaded_accuracy = 100.0 * OaFnMetric.Accuracy(
		reloaded.Forward(x), y
	)
	reloaded_generation = nlp._generate_sliding(
		reloaded,
		OaNlpSuiteSampler(recipe, 1),
		recipe,
		generation_units,
	)

	print(
		f"\nEvaluation: baseline={math.log(256):.4f} · "
		f"initial CE={initial_loss:.4f} · final CE={final_loss:.4f} · "
		f"accuracy={accuracy:.1f}%"
	)
	print(f"Generation: {generated!r}")
	assert initial_loss > 0.0
	assert math.isfinite(final_loss)
	assert abs(reloaded_accuracy - accuracy) < 0.5
	assert reloaded_generation == generated
	if full_gate:
		assert final_loss < initial_loss
		assert accuracy > 30.0
	print("✓ All checks passed")


if __name__ == "__main__":
	main()
